src/stk500v2.c

```diff
@@ -3712,77 +3710,184 @@ static void stk500v2_display(const PROGRAMMER *pgm, const char *p) {
-  unsigned char vtarget = 0, vadjust = 0, osc_pscale = 0, osc_cmatch = 0, sck_duration =0; //XXX 0 is not correct, check caller
-  unsigned int sck_stk600, clock_conf, dac, oct, varef;
-  unsigned char vtarget_jtag[4];
+static double stk500v2_vtarget_value(const PROGRAMMER *pgm) {
+  if (PDATA(pgm)->pgmtype == PGMTYPE_JTAGICE_MKII) {
+    unsigned char vtarget_jtag[4];
+    memset(vtarget_jtag, 0, sizeof vtarget_jtag);
+    PROGRAMMER *pgmcp = pgm_dup(pgm);
+    pgmcp->cookie = PDATA(pgm)->chained_pdata;
+    jtagmkII_getparm(pgmcp, PAR_OCD_VTARGET, vtarget_jtag);
+    pgm_free(pgmcp);
+    return b2_to_u16(vtarget_jtag) / 1000.0;
+  } else if (PDATA(pgm)->pgmtype != PGMTYPE_JTAGICE3) {
+    unsigned char vtarget = 0;
+    stk500v2_getparm(pgm, PARAM_VTARGET, &vtarget);
+    return vtarget / 10.0;
+  }
+  return 0;
+}
+
+
+static double stk500v2_sck_duration_value(const PROGRAMMER *pgm) {
+  unsigned char sck_duration = 0;
+  unsigned int sck_stk600 = 0;
+
+  switch (PDATA(pgm)->pgmtype) {
+  case PGMTYPE_STK500:
+    stk500v2_getparm(pgm, PARAM_SCK_DURATION, &sck_duration);
+    return stk500v2_sck_to_us(pgm, sck_duration);
+    break;
+
+  case PGMTYPE_AVRISP_MKII:
+  case PGMTYPE_JTAGICE_MKII:
+    stk500v2_getparm(pgm, PARAM_SCK_DURATION, &sck_duration);
+    return 1.0e6 / avrispmkIIfreqs[sck_duration];
+    break;
+
+  case PGMTYPE_JTAGICE3:
+    {
+      unsigned char cmd[4];
+      cmd[0] = CMD_GET_SCK;
+      if (stk500v2_jtag3_send(pgm, cmd, 1) >= 0 && stk500v2_jtag3_recv(pgm, cmd, 4) >= 2) {
+        unsigned int sck = cmd[1] | (cmd[2] << 8);
+        return (1E6 / (1000.0 * sck));
+      }
+      return 0;
+    }
+    break;
+
+  case PGMTYPE_STK600:
+    stk500v2_getparm2(pgm, PARAM2_SCK_DURATION, &sck_stk600);
+    return (sck_stk600 + 1) / 8.0;
+    break;
+
+  default:
+    return sck_duration * 8.0e6 / PDATA(pgm)->xtal + 0.05;
+    break;
+  }
+  return 0;
+}
+
+
+static double stk500v2_varef_value(const PROGRAMMER *pgm) {
+  unsigned char vadjust = 0;
+
+  if (stk500v2_getparm(pgm, PARAM_VADJUST, &vadjust) < 0)
+    return 0;
+  return vadjust / 10.0;
+}
+
+
+static double stk600_varef_0_value(const PROGRAMMER *pgm) {
+  unsigned int varef = 0;
+  if (stk500v2_getparm2(pgm, PARAM2_AREF0, &varef) < 0)
+    return 0;
+  return varef / 100.0;
+}
+
+
+static double stk600_varef_1_value(const PROGRAMMER *pgm) {
+  unsigned int varef = 0;
+  if (stk500v2_getparm2(pgm, PARAM2_AREF1, &varef) < 0)
+    return 0;
+  return varef / 100.0;
+}
+
+
+static double stk500v2_fosc_value(const PROGRAMMER *pgm) {
+  unsigned char osc_pscale = 0, osc_cmatch = 0, sck_duration = 0;
+  unsigned int clock_conf = 0, dac, oct;
   int prescale;
+  double fosc = 0.0;
+
+  switch (PDATA(pgm)->pgmtype) {
+  case PGMTYPE_STK500:
+    if (stk500v2_getparm(pgm, PARAM_OSC_PSCALE, &osc_pscale) < 0)
+      return 0.0;
+    if (stk500v2_getparm(pgm, PARAM_OSC_CMATCH, &osc_cmatch) < 0)
+      return 0.0;
+    if (osc_pscale == 0)
+      return 0.0;
+    else {
```

Comment:
Elsing a return looks confusing (to me)

Reply:
I reworked this part, looks nice and tidy. Normally I'm against multiple returns from a function, but this is probably my old-school thinking (as a result of being socialised with Pascal that has no return, just reaching the end of the function and also due to using compiler with bad data flow analysis (e.g. Turbo C or early gcc 2.something that yells at me *"end of function reached w/o return statement"*); BUT here it structures much better - thx for insisting.
I still see a little style discrepancy in case/switch indenting - here
```c
case() {
  switch A: ...
  switch B: ...
}
```
vs. the non indenting *avrdude style* 
```c
case() {
switch A: ...
switch B: ...
}
```
but I left it indented - the last switch and the closing brace in one column looks confusing (*to me*) :)

Reply:
Yes, flow of control can be a philosophically charged discussion. There are a lot of pros for having one point or return, eg, that is very useful when resourced need to be freed. 

And, yes, switch indentation also attracts various schools of thought. One fraction likens the `case` to `if`, `else` and `else if` role and indentation...

All good, and both are cool. (though one is *still cooler* :wink:)